dev/src/lobster/flat_gltf.h

```diff
@@ -0,0 +1,789 @@
+                             {0.0f, 0.0f, 0.0f, 1.0f}, // rotation
+                             {1.0f, 1.0f, 1.0f},       // scale
+                             children};
+        // not all glTF files have translation, rotation, and scale for each
+        // node here we check for the existence of each property
+        if (n->translation() && n->translation()->size() >= 3)
+          newNode.translation = {n->translation()->Get(0),
+                                 n->translation()->Get(1),
+                                 n->translation()->Get(2)};
+        if (n->rotation() && n->rotation()->size() >= 4)
+          newNode.rotation = {n->rotation()->Get(0), n->rotation()->Get(1),
+                              n->rotation()->Get(2), n->rotation()->Get(3)};
+        if (n->scale() && n->scale()->size() >= 3)
+          newNode.scale = {n->scale()->Get(0), n->scale()->Get(1),
+                           n->scale()->Get(2)};
+        scene.nodes.push_back(newNode);
+      }
+    }
+
+    _VERBOUT_ << "Total nodes:  " << scene.nodes.size() << std::endl
+              << std::endl;
+
+    if (info->animations()) {
+      for (size_t i = 0; i < info->animations()->size(); i++) {
+        auto an = info->animations()->Get(i);
+        scene.animations.resize(scene.animations.size() + 1);
+        scene.animations.back().name = an->name()->str();
+        for (size_t j = 0; j < an->samplers()->size(); j++) {
+          auto s = an->samplers()->Get(j);
+          _VERBOUT_ << "Input max:    " << accessors[s->input()].maxf[0]
+                    << std::endl;
+          // TODO: make this not crash the program
+          //_VERBOUT_ << "Input max:    " << accessors[ s->output() ].maxf[0] <<
+          //std::endl;
+        }
+        for (size_t j = 0; j < an->channels()->size(); j++) {
+          auto c = an->channels()->Get(i);
+        }
+      }
+    }
+
+    _VERBOUT_ << "Animations:   " << scene.animations.size() << std::endl
+              << std::endl;
+
+    // copy scenes[scene].nodes into scene.roots
+    auto info_scene_nodes = info->scenes()->Get(info->scene())->nodes();
+    for (size_t i = 0; i < info_scene_nodes->size(); i++)
+      scene.roots.push_back(info_scene_nodes->Get(i));
+
+    _VERBOUT_ << "Root nodes:   " << scene.roots.size() << std::endl
+              << std::endl;
+
+    return scene;
+  } catch (const std::exception &e) {
+    _ERROUT_ << "Caught exception: " << e.what() << std::endl << std::endl;
+  }
+  return newScene;
+}
+
+void SetSceneAnimation(Scene &scene, int anim, double animTime) {
+  SceneAnimation &animation = scene.animations[anim];
+}
+
+void DrawScene(Scene scene, float4x4 modelViewMat, float4x4 projMat) {
+  size_t ct = 0;
+  int currentMaterial = -1;
+  // https://shaharmike.com/cpp/lambdas-and-functions/
+  std::function<void(int, float4x4)> DrawTree = [&](int n, float4x4 m) {
+    ct++;
+    if (ct > MAX_SCENE_NODES)
+      throw std::out_of_range(
```

Comment:
If at all possible, have functions return a bool, error code or error string to indicate a problem, so the caller can decide if a bad file should mean the program terminates. Prefer to not use exception handling.. I know the rest of the codebase does for fatal runtime errors, but I am actually trying to use it as little as possible.

Reply:
Returning Value( 0 or 1 ) through the nfr. Hopefully that works.